pkg/ops/rawDiskGeneration.go

```diff
@@ -0,0 +1,700 @@
+	// Copy the grub.cfg
+	// Create dirs as needed
+	err = fsutils.MkdirAll(r.config.Fs, filepath.Join(tmpDirEfi, "EFI", "BOOT"), 0755)
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Str("target", tmpDirEfi).Msg("failed to create boot dir")
+		return "", err
+	}
+	flavor, err := sdkUtils.OSRelease("FLAVOR", filepath.Join(r.Source, "etc/kairos-release"))
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Msg("failed to get flavor")
+		return "", err
+	}
+	internal.Log.Logger.Debug().Str("flavor", flavor).Msg("got flavor")
+	model, err := sdkUtils.OSRelease("MODEL", filepath.Join(r.Source, "etc/kairos-release"))
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Msg("failed to get model")
+		return "", err
+	}
+	internal.Log.Logger.Debug().Str("model", model).Msg("got model")
+
+	if strings.Contains(flavor, "ubuntu") {
+		err = fsutils.MkdirAll(r.config.Fs, filepath.Join(tmpDirEfi, "EFI", "ubuntu"), 0755)
+		if err != nil {
+			internal.Log.Logger.Error().Err(err).Str("target", tmpDirEfi).Msg("failed to create ubuntu dir")
+			return "", err
+		}
+		err = r.config.Fs.WriteFile(filepath.Join(tmpDirEfi, "EFI", "ubuntu", "grub.cfg"), []byte(constants.GrubEfiRecovery), 0o644)
+		if err != nil {
+			internal.Log.Logger.Error().Err(err).Str("target", tmpDirEfi).Msg("failed to write grub.cfg")
+			return "", err
+		}
+	} else {
+		err = r.config.Fs.WriteFile(filepath.Join(tmpDirEfi, "EFI", "BOOT", "grub.cfg"), []byte(constants.GrubEfiRecovery), 0o644)
+		if err != nil {
+			internal.Log.Logger.Error().Err(err).Str("target", tmpDirEfi).Msg("failed to write grub.cfg")
+			return "", err
+		}
+	}
+
+	// Now search for the grubARCH.efi and copy it to the efi partition from the rootfs
+	if strings.Contains(strings.ToLower(flavor), "alpine") && strings.Contains(strings.ToLower(model), "rpi") {
+		internal.Log.Logger.Warn().Msg("Running on Alpine+RPI, not copying shim or grub.")
+	} else {
+		err = r.copyShimOrGrub(tmpDirEfi, "shim")
+		if err != nil {
+			internal.Log.Logger.Error().Err(err).Msg("failed to copy shim")
+			return "", err
+		}
+
+		err = r.copyShimOrGrub(tmpDirEfi, "grub")
+		if err != nil {
+			internal.Log.Logger.Error().Err(err).Msg("failed to copy grub")
+			return "", err
+		}
+	}
+
+	efiPartitionImage := v1.Image{
+		File:       filepath.Join(r.tmpDir, "efi.img"),
+		FS:         agentConstants.EfiFs,
+		Label:      agentConstants.EfiLabel,
+		Size:       20,
+		Source:     v1.NewDirSrc(tmpDirEfi),
+		MountPoint: tmpDirEfiMount,
+	}
+
+	// Deploy the source to the image
+	_, err = r.elemental.DeployImageNodirs(&efiPartitionImage, false)
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Str("source", r.Source).Interface("image", efiPartitionImage).Msg("failed to create efi image")
+		return "", err
+	}
+
+	return efiPartitionImage.File, nil
+}
+
+func (r *RawImage) Build() error {
+	var bootImagePath string
+	var err error
+
+	// If we dont have a tempdir, create one
+	if r.tmpDir == "" {
```

Comment:
I think I had this due to the methods being public VS private. 

So I though the Build as the entrypoint for the RawImage creation, which made sense. So we setup the tmpdir here and then its used everywhere.

But then I though that for the nvidia images it may be necessary to expose the methods as it also creates partitions or something and then merges.

But I think this is useless for now.

I will make the methods private and have only Build as entrypoint, which creates the tmpdir at the start and removes it at the end so its more straigthforward. If we need to reuse the functions in other places, we can revisit that down the line.

But I really like that idea, will probably add it anyway

Reply:
added!